File: Java Iteration/src/main/java/course/CourseJSONReader.java
package course;

import model.*;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.LinkedList;

public class CourseJSONReader {


    JSONArray lectureList = new JSONArray();

    public void readJson(LinkedList<Lecture>[] courseList, LinkedList<Teacher> teacherList, LinkedList<NonTechnicalElective> nonTechnicalList, LinkedList<TechnicalElective>[] technicalList, String semesterInput){


        JSONParser jsonParser = new JSONParser();

        try (
                FileReader reader = new FileReader("courses.json"))
        {
            //Read JSON file
            Object obj = jsonParser.parse(reader);

            lectureList = (JSONArray) obj;
            System.out.println(lectureList);

            //Iterate over employee array
            lectureList.forEach(
                    emp -> parseCourseObject( (JSONObject) emp ,courseList,teacherList,nonTechnicalList,technicalList, semesterInput)
            );

            for (int j = 0; j < 8 ; j++){
                for (int i = 1 ; i < courseList[j].size() ; i++) {
                    Lecture course =  courseList[j].get(i);
                    course.getName();
                    System.out.println(course.getName() +"-"+ course.getCode() +"-"+ course.getLecturer().getName() );
                }
            }
        } catch (
                FileNotFoundException e) {
            e.printStackTrace();
        } catch (
                IOException e) {
            e.printStackTrace();
        } catch (
                ParseException e) {
            e.printStackTrace();
        }

    }




    private void parseCourseObject(JSONObject employee, LinkedList<Lecture>[] courseList, LinkedList<Teacher> teacherList, LinkedList<NonTechnicalElective> nonTechnicalList, LinkedList<TechnicalElective>[] technicalList, String semestrInput)
    {
        if(employee==null){ return; }
        //Get course object within list
        JSONObject courseObject = (JSONObject) employee.get("Course");

        //Get course name
        String courseName = (String) courseObject.get("CourseName");
        System.out.print("Course name : " + courseName);

        //Get course credit
        int credit = Integer.parseInt((String) courseObject.get("Credit"));
        System.out.print(" -- Credit : " + credit);

        //Get course code
        String courseCode = (String) courseObject.get("CourseCode");
        System.out.print(" -- Course code : " + courseCode);

        //Get course lecturer
        String lecturer = (String) courseObject.get("Lecturer");
        System.out.print(" -- Lecturer : " + lecturer);
        Teacher teacher = findTeacher(teacherList ,lecturer);

        //Get course semestr
        int semestr = Integer.parseInt((String) courseObject.get("Semestr"));
        System.out.print(" -- Semestr : " + semestr);

        //Get course prerequisite
        String prerequisite = (String) courseObject.get("Prerequisite");
        System.out.println(" -- Prerequisite : " + prerequisite);

        String schedule = (String) courseObject.get("Schedule");
        System.out.println(" -- Schedule : " + schedule);

        Lecture course = null;
        if( !prerequisite.equals("") ) {
            for( int i = 0 ; i < courseList.length ; i++ ) {
                for( int j = 0 ; j < courseList[i].size() ; j++ ) {
                    if( courseList[i].get(j).getCode().equals(prerequisite) ){
                        course = courseList[i].get(j);
                        break;
                    }
                }
            }
        }




        if(courseCode.startsWith("NTE") || courseCode.startsWith("FTE") || courseCode.startsWith("UE")) {
            //Get if selected
            boolean isSelective = Boolean.parseBoolean((String) courseObject.get("Selective"));
            System.out.print(" -- Selective : " + courseCode);

            if(isSelective){
                nonTechnicalList.add(new NonTechnicalElective(courseName,courseCode,teacher,credit,Integer.parseInt((String)courseObject.get("Quota")),schedule,""));
            }

            else{
                courseList[semestr-1].add(new NonTechnicalElective(courseName,courseCode,teacher,credit,Integer.parseInt((String)courseObject.get("Quota")),schedule,""));
            }
        }
        else if(courseCode.startsWith("TE")) {
            //Get if selected
            boolean isSelective = Boolean.parseBoolean((String) courseObject.get("Selective"));
            System.out.print(" -- Selective : " + courseCode);
            if(isSelective){
                if( semestr==7 ) {
                    technicalList[0].add(new TechnicalElective(courseName,courseCode,teacher,credit,Integer.parseInt((String)courseObject.get("Quota")),course,schedule,""));
                }
                else {
                    technicalList[1].add(new TechnicalElective(courseName,courseCode,teacher,credit,Integer.parseInt((String)courseObject.get("Quota")),course ,schedule,""));
                }
            }
            else {
                courseList[semestr-1].add(new TechnicalElective(courseName,courseCode,teacher,credit,Integer.parseInt((String)courseObject.get("Quota")),course,schedule,"" ));
            }
        }
        else {
            Course crs = new Course(courseName,courseCode,teacher,credit,credit,course,semestr,schedule,"");
            courseList[semestr-1].add(new Course(courseName,courseCode,teacher,credit,credit,course,--semestr,schedule,"")) ;
        }

    }

    private Teacher findTeacher(LinkedList<Teacher> teacherList, String lecturerName) {

        for ( int i = 0 ; i < teacherList.size() ; i++ ) {

            if( teacherList.get(i).getName().equals(lecturerName) ) {
                return teacherList.get(i);
            }

        }

        return new Teacher("Computer Engineering",lecturerName,""+Counter.id++);
    }
}

File: Python Iteration/CourseJSONReader.py
import json
import traceback

from Course import Course
from NE import NE
from TE import TE


class CourseJSONReader:

    def readJson(self):

        try:
            with open('courses.json') as json_file:
                data = json.load(json_file)['employees']
                technicalList= [[],[]]
                nonTechnicalList= []
                courseList = [[], [], [], [], [], [], [], []]

                for i in data:
                    """print(i["CourseName"])"""
                    """{
                        "CourseName": "Calculus 1",
                        "CourseCode": "MATH1001",
                        "Credit": "6",
                        "Lecturer": "Mustafa Cem \u00c7elik",
                        "Semestr": "1",
                        "Prerequisite": "",
                        "Schedule": "Sal\u0131 08:30-10:20 Per\u015fembe 08:30-10:20"
                    },"""
                    """print( i["CourseCode"])
                    print( i["CourseCode"].startswith("TE"))"""
                    if i["CourseCode"].startswith("TE"):
                        # name, code, credit,lecturer,semester, prerequisite, schedule,quota):

                        # name, code,lecturer, credit,schedule, semester, prerequisite
                        te_course = TE(i["CourseName"], i["CourseCode"], i["Credit"], i["Lecturer"], i["Semestr"], i["Prerequisite"], i["Schedule"], i["Quota"])

                        if int(te_course.semester)==7:
                            technicalList[0].append(te_course)
                        else:
                            technicalList[1].append(te_course)


                    elif i["CourseCode"].startswith("NTE"):
                        """{
                            "CourseName": "The Culture of Radiation Safety",
                            "CourseCode": "NTE1003",
                            "Credit": "3",
                            "Lecturer": "Deniz \u00d6ner",
                            "Semestr": "2",
                            "Prerequisite": "",
                            "Schedule": "Sal\u0131 17:00-19:00",
                            "Quota": "20"
                        },"""


                        nte_course = NE(i["CourseName"], i["CourseCode"], i["Credit"], i["Lecturer"], i["Semestr"], i["Schedule"], i["Quota"])
                        nonTechnicalList.append(nte_course)
                    else:

                        """{
                            "CourseName": "Calculus 1",
                            "CourseCode": "MATH1001",
                            "Credit": "6",
                            "Lecturer": "Mustafa Cem \u00c7elik",
                            "Semestr": "1",
                            "Prerequisite": "",
                            "Schedule": "Sal\u0131 08:30-10:20 Per\u015fembe 08:30-10:20"
                        },"""


                        semester = int(i["Semestr"])-1
                        #                   name,           code,           lecturer,      credit,      schedule, semester,     prerequisite
                        course = Course(i["CourseName"], i["CourseCode"],i["Lecturer"], i["Credit"], i["Schedule"],semester, i["Prerequisite"] )
                        if course.code in ["CSE1141","CSE1142","CSE1142"]:
                            course.passed = True
                        courseList[semester].append(course)


            return courseList,nonTechnicalList,technicalList


        except FileNotFoundError:
            traceback.print_exc()
        except IOError:
            traceback.print_exc()


